Bidirectional Orders Fulfillment Sync between WooCommerce and Square Orders

faisal-alvi opened the pull request:
### All Submissions:

<!-- Mark completed items with an [x] -->
* [ ] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:

Implements bidirectional fulfillment synchronization between WooCommerce and Square orders, enhancing the existing order sync functionality.

#### Key Implementation Details:

- Adds fulfillment details when creating Square orders from WooCommerce orders
- Syncs fulfillment status changes from Square back to WooCommerce orders
- Maintains order status consistency across both platforms
- Leverages Square's OrderFulfillment API for reliable state management

#### Important Notes:

- Only orders processed through Square Payment Gateway are synced to Square
- A polling system is used to fetch Square orders every 15 minutes by default (this interval can be modified using a filter; `wc_square_order_polling_interval_seconds`).
- Webhooks are not used, as they require merchants to create an application in their Square dashboard; a process that may not be straightforward for most users.
- Only the existing Woo orders are eligible for the syncing of fulfillment updates from Square. No new Woo orders are created.
- Only Square order updates that mark an order as either COMPLETED or CANCELLED are synced back to WooCommerce, as those are the only statuses we are targeting for now. When a fulfillment is completed, the corresponding Square order is marked as COMPLETED, which triggers the sync. Similarly, we also track if the order is marked as CANCELLED.
- For refund scenarios, Square creates a new order while WooCommerce deals with the same order, as both platforms have different strategies, refund scenarios are skipped in this initial implementation.
- No new settings added as order sync was partially implemented for the Woo→Square direction already
- No new database tables have been added, and no new button has been introduced to trigger manual sync in this implementation
- This implementation establishes a foundational integration between the platforms, with opportunities for future enhancements

#### Note on Sync Loop Prevention:

The current implementation intentionally avoids syncing WooCommerce order updates back to Square to prevent infinite sync loops. Here’s the potential issue:

1. A WooCommerce order is updated
2. The update syncs to Square
3. The Square polling system detects the change
4. The change syncs back to WooCommerce
5. Steps 1–4 repeat indefinitely

While it’s possible to implement a solution using metadata flags, such approaches may introduce edge cases. It’s important that we carefully evaluate the implications before heading in that direction.

As a result, this implementation includes:
1. Fulfillment data is included when creating a new Square order
2. Fulfillment updates made in Square are synced back to the same WooCommerce order

Closes https://linear.app/a8c/issue/SQUARE-126/implement-two-way-order-sync.
Closes https://linear.app/a8c/issue/SQUARE-98/created-orders-do-not-appear-in-the-merchants-square-dashboard

### Steps to test the changes in this Pull Request:

1. **Test WooCommerce to Square Sync:**
   - Create a new order in WooCommerce using Square Payment Gateway
   - Verify order appears in Square Dashboard with correct fulfillment details
   - Check fulfillment options are visible at the Square order and match with WooCommerce

2. **Test Square to WooCommerce Sync:**
   - Update fulfillment status in Square Dashboard (e.g., mark as delivered/shipped)
   - Verify Square order status is now "completed"
   - Verify WooCommerce order status updates accordingly (this may require waiting for 15 minutes or manually running `wc_square_sync_orders` event; run from the `/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending` page)
   - Check WooCommerce order is set to "completed"

3. **Test with Different Payment Methods:**
   - Check the same process for Credit Card, Cash App, and Digital Wallet Payments.
   - Check order with Gift Card.

4. **Test Other Cases:**
   - Confirm Capture Order functionality remains unaffected and continues to work as expected
   - Confirm proper error handling for failed syncs
   - Check sync behavior with cancelled/refunded orders

### Changelog entry

> Add - Order fulfillment sync between WooCommerce and Square orders.

faisal-alvi committed:
Add order fulfillment handling for new Woo orders

faisal-alvi committed:
Merge branch 'trunk' into feat/bidirect-order-sync-woo-to-square

faisal-alvi committed:
Add Order Polling class for syncing orders from Square to WooCommerce

faisal-alvi committed:
Add custom cron schedule for Square order polling

faisal-alvi committed:
Add functionality to schedule polling for Square orders

faisal-alvi committed:
Add polling functionality for Square orders

faisal-alvi committed:
Add method to fetch recent Square orders in Order_Polling class

faisal-alvi committed:
Add method to search Square orders since a specific time in Order_Polling class

faisal-alvi committed:
Process Square orders and update WooCommerce orders

faisal-alvi committed:
Retrieve last polling time

faisal-alvi committed:
Update last polling time

faisal-alvi committed:
Add Order_Importer class for syncing Square orders with WooCommerce

faisal-alvi committed:
Update fulfillment data for syncing orders

faisal-alvi committed:
Find existing WooCommerce orders by Square order ID

faisal-alvi committed:
Add Order_Mapper class to handle mapping between Square and WooCommerce orders

faisal-alvi committed:
Add method to search for Square orders with filters and pagination in API class

faisal-alvi committed:
Add Order_Sync handler to Plugin class

faisal-alvi committed:
Add Order_Sync class to handle syncing WooCommerce orders with Square

faisal-alvi committed:
Use direct database query for improved performance and reliability

faisal-alvi committed:
Add early return in `maybe_sync_update_to_square()` for future updates

faisal-alvi committed:
Check for response errors instead of transaction approval; as now we order is still open because fulfillments may not done yet

faisal-alvi committed:
Pass fulfillment data only for new orders

faisal-alvi committed:
phpcs fixes for API.php
-----------------------------------------------------------------------------------------------
FOUND 12 ERRORS AND 3 WARNINGS AFFECTING 15 LINES
-----------------------------------------------------------------------------------------------
 736 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 741 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 744 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 747 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 750 | WARNING | Equals sign not aligned with surrounding assignments; expected 3 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 758 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 760 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 765 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 767 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 770 | WARNING | Equals sign not aligned with surrounding assignments; expected 3 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 771 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 775 | ERROR | Using short ternaries is not allowed as they are rarely used correctly (Universal.Operators.DisallowShortTernary.Found)
 779 | WARNING | Equals sign not aligned with surrounding assignments; expected 9 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 785 | ERROR | Blank line found after control structure (WordPress.WhiteSpace.ControlStructureSpacing.BlankLineAfterEnd)
 786 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
phpcs fixes for Order_Mapper.php,

-----------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 2 LINES
-----------------------------------------------------------------------------------------------
 51 | WARNING | Not using strict comparison for in_array; supply true for $strict argument. (WordPress.PHP.StrictInArray.MissingTrueStrict)
 60 | WARNING | Not using strict comparison for in_array; supply true for $strict argument. (WordPress.PHP.StrictInArray.MissingTrueStrict)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
Remove unused "future" method `maybe_sync_update_to_square()` for now

faisal-alvi committed:
phpcs fixes for Order_Importer.php

-----------------------------------------------------------------------------------------------
FOUND 19 ERRORS AND 4 WARNINGS AFFECTING 23 LINES
-----------------------------------------------------------------------------------------------
  51 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
  51 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  52 | ERROR | A function call to __() with texts containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. (WordPress.WP.I18n.MissingTranslatorsComment)
  52 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  53 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  54 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
  56 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
  56 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  57 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  58 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  59 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  60 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  61 | ERROR | Only one argument is allowed per line in a multi-line function call (PEAR.Functions.FunctionCallSignature.MultipleArguments)
  61 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
  66 | WARNING | Not using strict comparison for in_array; supply true for $strict argument. (WordPress.PHP.StrictInArray.MissingTrueStrict)
  70 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  78 | WARNING | Equals sign not aligned with surrounding assignments; expected 7 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 104 | ERROR | There should be a comma after the last array item in a multi-line array. (NormalizedArrays.Arrays.CommaAfterLast.MissingMultiLine)
 115 | ERROR | There should be a comma after the last array item in a multi-line array. (NormalizedArrays.Arrays.CommaAfterLast.MissingMultiLine)
 135 | WARNING | Equals sign not aligned with surrounding assignments; expected 2 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 162 | ERROR | A function call to __() with texts containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. (WordPress.WP.I18n.MissingTranslatorsComment)
 171 | WARNING | Equals sign not aligned with surrounding assignments; expected 5 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 183 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
phpcs fixes for Order_Polling.php

-----------------------------------------------------------------------------------------------
FOUND 47 ERRORS AND 10 WARNINGS AFFECTING 57 LINES
-----------------------------------------------------------------------------------------------
  37 | WARNING | Detected changing of cron_schedules, but could not detect the interval value. (WordPress.WP.CronInterval.ChangeDetected)
 109 | ERROR | String "Square order polling completed." does not require double quotes; use single quotes instead (Squiz.Strings.DoubleQuoteUsage.NotRequired)
 157 | WARNING | Equals sign not aligned with surrounding assignments; expected 2 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 158 | WARNING | Equals sign not aligned with surrounding assignments; expected 6 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 169 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
 169 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 170 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 171 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 173 | ERROR | Only one argument is allowed per line in a multi-line function call (PEAR.Functions.FunctionCallSignature.MultipleArguments)
 173 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
 178 | WARNING | Stand-alone post-increment statement found. Use pre-increment instead: ++$batch_count. (Universal.Operators.DisallowStandalonePostIncrementDecrement.PostIncrementFound)
 182 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
 182 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 183 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 184 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 185 | ERROR | Only one argument is allowed per line in a multi-line function call (PEAR.Functions.FunctionCallSignature.MultipleArguments)
 185 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
 208 | WARNING | Equals sign not aligned with surrounding assignments; expected 3 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 209 | WARNING | Equals sign not aligned with surrounding assignments; expected 3 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 210 | WARNING | Equals sign not aligned with surrounding assignments; expected 5 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 224 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
 224 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 225 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 226 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 229 | ERROR | Only one argument is allowed per line in a multi-line function call (PEAR.Functions.FunctionCallSignature.MultipleArguments)
 229 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
 230 | WARNING | Stand-alone post-increment statement found. Use pre-increment instead: ++$updated_count. (Universal.Operators.DisallowStandalonePostIncrementDecrement.PostIncrementFound)
 232 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
 232 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 233 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 234 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 237 | ERROR | Only one argument is allowed per line in a multi-line function call (PEAR.Functions.FunctionCallSignature.MultipleArguments)
 237 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
 238 | WARNING | Stand-alone post-increment statement found. Use pre-increment instead: ++$skipped_count. (Universal.Operators.DisallowStandalonePostIncrementDecrement.PostIncrementFound)
 240 | ERROR | Blank line found after control structure (WordPress.WhiteSpace.ControlStructureSpacing.BlankLineAfterEnd)
 243 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
 243 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 244 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 245 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 246 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 247 | ERROR | Only one argument is allowed per line in a multi-line function call (PEAR.Functions.FunctionCallSignature.MultipleArguments)
 247 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
 248 | WARNING | Stand-alone post-increment statement found. Use pre-increment instead: ++$error_count. (Universal.Operators.DisallowStandalonePostIncrementDecrement.PostIncrementFound)
 251 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
 251 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 252 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 253 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 254 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 255 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
 260 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
 260 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 261 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 262 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 264 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 265 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 266 | ERROR | Only one argument is allowed per line in a multi-line function call (PEAR.Functions.FunctionCallSignature.MultipleArguments)
 266 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
-----------------------------------------------------------------------------------------------

Sidsector9 commented on includes/Handlers/Order_Sync.php:
```diff
@@ -0,0 +1,34 @@
+class Order_Sync {
+	public function __construct() {
+		// Initialize hooks.
+		$this->init();
```
`self::init()`

faisal-alvi replied:
Thanks for the feedback. I've removed the `static` keyword instead.

faisal-alvi committed:
Update `init` from static to instance method

dkotter reviewed (commented):
Overall this looks good to me though I didn't have time to do a super thorough review, pinging in @Sidsector9 to do a deeper review.

Will definitely want to ensure we have tests in place prior to releasing this

Sidsector9 committed:
Merge branch 'trunk' into feat/bidirect-order-sync-woo-to-square

Sidsector9 commented on includes/Gateway/API.php:
```diff
@@ -716,6 +716,78 @@ public function retrieve_gift_card_by_gan( $gan = '' ) {
+	 * @return array Array containing orders and cursor.
+	 * @throws \Exception
+	 */
+	public function search_orders( $location_ids, $start_time, $limit = 100, $cursor = '' ) {
```
```suggestion
	public function search_orders( $location_ids = array(), $start_time, $limit = 100, $cursor = '' ) {
```

Sidsector9 commented on includes/Gateway/API.php:
```diff
@@ -716,6 +716,78 @@ public function retrieve_gift_card_by_gan( $gan = '' ) {
+	 *
+	 * @since x.x.x
+	 * @param array $location_ids Array of location IDs.
+	 * @param string $start_time Start time for the search.
```
```suggestion
	 * @param string $start_time Start time in the RFC 3339 format for the search.
```

Sidsector9 commented on includes/Gateway/API.php:
```diff
@@ -716,6 +716,78 @@ public function retrieve_gift_card_by_gan( $gan = '' ) {
+			wc_square()->log( 'SearchOrders API error: ' . $e->getMessage(), 'error' );
+			throw $e;
+		}
+	}
```
Let us do this similar to how we implemented other APIs so that the architecture remains consistent across.

This logic can go under `Gateway\API\Requests\Orders.php`.

The response can be set using a response handler, for example:

```php
$this->set_response_handler( \WooCommerce\Square\API\Response::class );
```

And the request can be performed using:

```php
$response = $this->perform_request( $request );
```

The response can then be checked for:

```php
$response->get_data() instanceof \Square\Models\SearchOrdersResponse
```

Sidsector9 commented on includes/Gateway/API/Requests/Orders.php:
```diff
@@ -169,6 +175,91 @@ function( $line_item ) {
+
+		// Create comprehensive fulfillment object.
+		$fulfillment = new \Square\Models\OrderFulfillment( 'PROPOSED' );
+		$fulfillment->setUid( 'woo_order_fulfillment_' . $order->get_id() );
```
Throughout the plugin, we use `wc_square()->get_idempotency_key()` to set UUID. Any specific reason we are choosing to do it differently here?

Sidsector9 commented on includes/Gateway/API/Requests/Orders.php:
```diff
@@ -169,6 +175,91 @@ function( $line_item ) {
+
+		// Determine fulfillment type based on shipping method.
+		$shipping_methods = $order->get_shipping_methods();
+		$fulfillment_type = empty( $shipping_methods ) ? 'PICKUP' : 'SHIPMENT';
```
Minor nitpick, allows for easier jump-to navigation.
```php
use Square\Models\FulfillmentType;
$fulfillment_type = empty( $shipping_methods ) ? FulfillmentType::PICKUP : FulfillmentType::SHIPMENT;
```

Sidsector9 replied:
Will this set the Fulfillment type to `PICKUP` for Virtual Orders and Downloadable content that don't require shipping?

faisal-alvi replied:
Yes, you’re right - assigning `PICKUP` fulfillment to virtual/downloadable products isn’t semantically ideal.

Currently, we’re using `PICKUP` because virtual products don’t require shipping, and this approach ensures such orders are recognized and visible in Square’s system. It’s not a perfect fit, but it maintains functional consistency with our integration requirements.

While skipping fulfillment for virtual products could be more accurate, the current Phase 1 scope requires fulfillment data to be passed for all orders. Additionally, Square mandates fulfillment for orders to be listed in their system (see [Linear#98](https://linear.app/a8c/issue/SQUARE-98/created-orders-do-not-appear-in-the-merchants-square-dashboard)).

So for now, I suggest we retain the current behavior. The use of `PICKUP` for virtual items, though imperfect, ensures compatibility. It might be worth flagging this with the Square team.

I’ll reach out to the Square team to confirm the best approach here and also request support for a more appropriate `DIGITAL` fulfillment type.

Sidsector9 replied:
Thanks for the info. I think we should brainstorm and think of the long-term implications of using PICKUP for virtual orders. @iamdharmesh @jeffpaul any thoughts on this?

faisal-alvi replied:
I followed up with the Square team ([on slack](https://fueled.slack.com/archives/C05REAQUK28/p1753794822695979)) and got confirmation from Max@Square:

> Unfortunately, releasing the DIGITAL fulfillment type on the public API is not on this year’s roadmap. I’m working to ensure this request is acknowledged by the product team.
So if using order fulfillments, the PICKUP type will be the best option for now.

So as things stand, continuing with `PICKUP` for virtual/downloadable orders remains the most compatible option for visibility in Square. We can flag this for future revision once Square expands fulfillment support.

Sidsector9 commented on includes/Gateway/API/Requests/Orders.php:
```diff
@@ -169,6 +175,91 @@ function( $line_item ) {
+
+			// Add recipient information.
+			$recipient = new \Square\Models\OrderFulfillmentRecipient();
+			$recipient->setDisplayName( trim( $order->get_shipping_first_name() . ' ' . $order->get_shipping_last_name() ) );
```
```suggestion
			$recipient->setDisplayName( trim( $order->get_formatted_shipping_full_name() ) );
```

Sidsector9 commented on includes/Handlers/Order_Sync.php:
```diff
@@ -0,0 +1,34 @@
+	 *
+	 * @since x.x.x
+	 */
+	public function init() {
```
The `init()` method is redundant if we're going to call it inside `__construct()`. As per best practices, we don't initialize classes and register hooks inside `__construct()`. I'd suggest to call `init()` on the instance of `Order_Sync` instead of calling it inside the constructor.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+					)
+				);
+			} else {
+				wc_square()->log(
```
Do we have plans in next phases to create a sync status page, like how we have for products? If yes, we should add a `// @todo: move this to a separate order sync status page`

faisal-alvi replied:
Sure, it’s not a required item, but we can definitely consider implementing it in a future phase. For now, adding a todo comment sounds good and won’t hurt, happy to include that.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+					$wc_order->set_total( $square_total_amount );
+					$updates_made[] = sprintf( 'Total: %s → %s', $current_total, $square_total_amount );
+				}
+			}
```
Can we use the `Money_Utility` class for this purpose?

```php
$currency             = $square_total->getCurrency();
$square_total_amount = Money_Utility::cents_to_float( $square_total->getAmount(), $currency );
$current_total        = (float) $wc_order->get_total();
$decimals             = Money_Utility::get_currency_decimals( $currency );

$square_total_rounded  = round( $square_total_amount, $decimals );
$current_total_rounded = round( $current_total, $decimals );

if ( $square_total_rounded !== $current_total_rounded ) {
	// Only update total if order is not paid yet or if it's a refund scenario.
	if ( ! $wc_order->is_paid() || $square_total_rounded < $current_total_rounded ) {
		$wc_order->set_total( $square_total_rounded );
		$updates_made[] = sprintf( 'Total: %s → %s', $current_total, $square_total_rounded );
	}
}
```

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+		}
+
+		// 3. Update order total if changed (with safety checks).
+		if ( $square_total ) {
```
This will always evaluate to `true` as `getTotalMoney()` always returns an instance of `Square\Models\Money`

faisal-alvi replied:
You're absolutely correct! The condition will always evaluate to `true`, we can check the actual amount value instead.

Sidsector9 replied:
@faisal-alvi, your previous implementation was correct. I didn't notice it return `?Money` and not `Money`, which means it returns a nullable type.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+						$wc_order->get_id()
+					),
+					'sync'
+				);
```
Just to confirm, should we `return` here if status change is not allowed?

faisal-alvi replied:
Good question! No, we should not return early if the status change is not allowed.

The current logic allows for partial updates - even if status change is blocked, other updates might still be valid:
- Fulfillment status updates
- Order total adjustments  
- Timestamp updates
- Fulfillment data sync

For example, a status change might be blocked (e.g., can't go from 'completed' to 'processing'), but fulfillment status updates are still valid and important.

Getting partial updates is better than getting no updates at all, so the current approach of logging the blocked change but continuing with other updates is the right pattern.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+	 * @param \WC_Order $wc_order WooCommerce order.
+	 * @param \Square\Models\Order $square_order Square order.
+	 */
+	private function update_fulfillment_data( $wc_order, $square_order ) {
```
This is a minor nitpick based on the observations in the plugin codebase. If we have a function that may or may not do updates, we prefix them with `maybe_`. Since this method can return early without making any updates, we can name it as `maybe_update_fulfillment_data`.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+						// translators: %s is the tracking number.
+						$note .= sprintf( __( ' Tracking: %s', 'woocommerce-square' ), $tracking_number );
+					}
+					$wc_order->add_order_note( $note );
```
As this is inside a loop, this will add multiple order notes for a Woo order.
If we really want to add a note for each Square Order Line Item, then should we update the order note to `Order Line Item fulfillment completed in Square.`?

faisal-alvi replied:
You're right that this is inside a loop, but there's an important distinction here:

We're adding one fulfillment per order that targets ALL items in that order. We're not using Square's beta `line_item_application` field to target specific items per fulfillment.

So even if there are multiple fulfillments, each represents a distinct fulfillment action (e.g., different tracking numbers) - not individual line item fulfillments.

The current note "Order fulfillment completed in Square" is accurate because:
- Each fulfillment covers all items in the order
- Each fulfillment has unique details (tracking number, etc.) that should be recorded
- We're not doing line-item-specific fulfillments

If we were using `line_item_application` to target specific items, then "Order Line Item fulfillment completed" would be more appropriate. But since we're targeting all items per fulfillment, the current wording is correct.

We can consider implementing that in future updates, when we begin supporting the sync of orders originally created in Square, as those may include line-item-level fulfillment.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+
+			// Handle pickup details
+			$pickup_details = $fulfillment->getPickupDetails();
+			if ( $pickup_details ) {
```
Redundant check as it will always evaluate to `true` as `getPickupDetails()` returns an instance of `Square\Models\FulfillmentPickupDetails`.

faisal-alvi replied:
Actually, `getPickupDetails()` can return null if the shipment details haven’t been set, so the check is necessary to avoid potential errors.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+
+			// Handle shipment details
+			$shipment_details = $fulfillment->getShipmentDetails();
+			if ( $shipment_details ) {
```
Redundant check as it will always evaluate to `true` as `getShipmentDetails()` returns an instance of `Square\Models\FulfillmentShipmentDetails`.

faisal-alvi replied:
Actually, `getShipmentDetails()` can return null if the shipment details haven’t been set, so the check is necessary to avoid potential errors.

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+			$wpdb->prepare(
+				"
+				SELECT order_id, meta_key
+				FROM {$wpdb->prefix}wc_orders_meta
```
This assumes the merchant is already using `HPOS`. Should we consider scenarios where the merchant is still using `wp_post` and `wp_postmeta` tables?

Sidsector9 commented on includes/Sync/Order_Importer.php:
```diff
@@ -0,0 +1,232 @@
+	 * @param string $square_order_id Square order ID.
+	 * @return \WC_Order|false
+	 */
+	public function find_existing_wc_order_by_square_id( $square_order_id ) {
```
```suggestion
	public function find_existing_wc_order_by_square_order_id( $square_order_id ) {
```

Sidsector9 commented on includes/Sync/Order_Polling.php:
```diff
@@ -0,0 +1,334 @@
+	 * @since x.x.x
+	 */
+	public function schedule_polling() {
+		wp_schedule_event( time(), 'square_order_polling', self::CRON_HOOK );
```
Any reason why we're not using Action Scheduler the way we're using to sync Products?

faisal-alvi replied:
You're right! Using Action Scheduler would be much better than WordPress cron for this use case. Let me do that. Thanks!

Sidsector9 commented on includes/Sync/Order_Polling.php:
```diff
@@ -0,0 +1,334 @@
+
+			do {
+				// Use the API's search_orders method with cursor
+				$response = $api->search_orders( array( $location_id ), $since_time, 100, $cursor );
```
Do you think we should take the Shifting Window Problem into consideration here?

Since this loop paginates using `since_time` and the cursor, but Square Orders can still be updated during sync, there's a risk that the result set could shift mid-pagination. Specifically, an order that was already processed could get updated from the Square side during the same sync, causing it to move to a later page and be fetched again – leading to duplicate processing.

Should we consider capturing a sync_start_time at the beginning and bounding the query with `updated_at > last_poll_time AND updated_at <= sync_start_time` to avoid both duplication and missing edge cases?

Curious to hear your thoughts!

faisal-alvi replied:
Great question about the Shifting Window Problem! You're absolutely right that pagination with time-based queries can cause duplicate processing.

I considered the bounded time window approach but it got complex and heavy in terms of performance, so went with a lightweight solution instead: https://github.com/woocommerce/woocommerce-square/pull/382/commits/0266a469a880d5205bb94fae1d2083a52822bdc1

Reasons:
- Orders are overwritten (not duplicated), so impact is minimal
- 15-minute polling + 100 orders per batch = very rare occurrence
- YAGNI principle - premature optimization

Sidsector9 replied:
@faisal-alvi this implements dedpulication, but it is still susceptible to the Shifting Window problem. Consider this example:

- You're querying Square with `updated_at > 10:00 AM`
- The sync starts at 10:10 AM (this is your `sync_start_time`).
- Orders are fetched in pages of 3, sorted by `updated_at ASC`.
- Plugin limits itself to `max_batches = 10`

Square orders (sorted by `updated_at`):

| Order ID | updated\_at |
| -------- | ----------- |
| #A       | 10:01 AM    |
| #B       | 10:02 AM    |
| #C       | 10:03 AM    |
| #D       | 10:04 AM    |
| #E       | 10:05 AM    |
| #F       | 10:06 AM    |
| #G       | 10:07 AM    |

#### Sync timeline

| Time         | What Happens                                                                    |
| ------------ | ------------------------------------------------------------------------------- |
| 10:10:01     | Sync starts, captures `sync_start_time = 10:10:01`.                             |
| 10:10:02     | Plugin queries Page 1 with `since=10:00`, gets `[#A, #B, #C]`.                  |
| 10:10:03     | Plugin starts processing `#A`.                                                  |
| 10:10:04     | Plugin processes `#B`.                                                          |
| **10:10:05** | **Square updates `#B`, sets `updated_at = 10:10:08 AM`.**                          |
| 10:10:06     | Plugin processes `#C`.                                                          |
| 10:10:07     | Plugin queries Page 2 using cursor (next page after `#C`). Gets `[#D, #E, #F]`. |
| 10:10:08     | Plugin processes `#D`.                                                          |
| 10:10:09     | Plugin processes `#E`.                                                          |
| 10:10:10     | Plugin processes `#F`.                                                          |
| 10:10:11     | Plugin queries Page 3, gets `[#G]`.                                             |
| 10:10:12     | Plugin processes `#G`.                                                          |
| 10:10:13     | Plugin ends pagination, updates `last_polling_time = 10:10:01`.                 |

Now, when you use `last_polling_time` as the new `sync_start_time` as `10:10:01 AM`, even though `#B` was updated at `10:08 AM`, it never fetches `#B` again because you're fetching all orders after `10:10:01 AM`.

faisal-alvi replied:
Thanks for the explanation @Sidsector9!

I’ve reverted to the lightweight solution and updated the logic to include the end time in the search order query.

You can [click here](https://github.com/woocommerce/woocommerce-square/pull/382/files/36d0ad9a189feb80491a3b21996119bf5816f4a0..e225bc77957c178117e39c2e70ecc34cf260b9a4) to review the latest changes.

iamdharmesh commented on includes/Gateway/API/Requests/Orders.php:
```diff
@@ -114,10 +114,16 @@ public function set_update_order_data( \WC_Order $order, \Square\Models\Order $s
 	 *
 	 * @param \WC_Order            $order        WooCommerce order object.
 	 * @param \Square\Models\Order $square_order Square order object.
+	 * @param string $order_request_type The type of order request.
```
Minor nitpick:
```suggestion
	 * @param \WC_Order            $order              WooCommerce order object.
	 * @param \Square\Models\Order $square_order       Square order object.
	 * @param string               $order_request_type The type of order request.
```

faisal-alvi committed:
Merge branch 'trunk' into feat/bidirect-order-sync-woo-to-square

faisal-alvi committed:
Enhance doc for `search_orders()` & set default value for `location_ids` param

faisal-alvi committed:
Add Search_Orders response class for handling API order search results

faisal-alvi committed:
Update `search_orders()` to utilize new `Orders` request class for improved structure and maintainability

faisal-alvi committed:
Use idempotency key for improved consistency

faisal-alvi committed:
use FulfillmentType constants for improved clarity and maintainability

faisal-alvi committed:
use formatted full name method for improved accuracy

faisal-alvi committed:
Call `init()` on the instance of `Order_Sync` instead of calling it inside the constructor

faisal-alvi committed:
Add TODO comment for potential order sync status page implementation

faisal-alvi committed:
use the `Money_Utility` class in order total calculation for improved accuracy and rounding consistency

faisal-alvi committed:
Check amount before calculation

faisal-alvi committed:
Allow specific transitions and early returns via filter

faisal-alvi committed:
Rename `update_fulfillment_data` to `maybe_update_fulfillment_data` for improved clarity in fulfillment update logic

faisal-alvi committed:
Remove always returning true conditions

faisal-alvi committed:
Implement HPOS compatibility in order retrieval logic

faisal-alvi committed:
Rename method `find_existing_wc_order_by_square_id` to `find_existing_wc_order_by_square_order_id` for improved clarity in order retrieval logic

faisal-alvi committed:
use Action Scheduler instead of cron for order polling

faisal-alvi committed:
Lighweight solution for "Shifting Window Problem"

faisal-alvi committed:
Minor doc formatting for improved readability

faisal-alvi committed:
Handle Race Condition Scenario: Capture sync start time and updating last polling time accordingly,

    // 1. Get last polling time (e.g., 10:00 AM)
    $last_polling_time = $this->get_last_polling_time();

    // 2. Fetch orders since 10:00 AM (takes 2-3 minutes to process)
    $orders = $this->fetch_recent_square_orders();

    // 3. Process orders (takes another 2-3 minutes)
    $this->process_square_orders( $orders );

    // 4. Update polling time to NOW (e.g., 10:05 AM) ❌ PROBLEM!
    $this->update_last_polling_time();

faisal-alvi committed:
phpcs fixes in several files,
FILE: includes/Gateway/API/Responses/Search_Orders.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 55 | ERROR | The closing brace for the class must go on the next line after the body (PSR2.Classes.ClassDeclaration.CloseBraceAfterBody)
-----------------------------------------------------------------------------------------------

FILE: includes/Sync/Order_Importer.php
-----------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 4 WARNINGS AFFECTING 4 LINES
-----------------------------------------------------------------------------------------------
 86 | WARNING | Equals sign not aligned with surrounding assignments; expected 12 spaces but found 13 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 87 | WARNING | Equals sign not aligned with surrounding assignments; expected 1 space but found 2 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 88 | WARNING | Equals sign not aligned with surrounding assignments; expected 7 spaces but found 8 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 89 | WARNING | Equals sign not aligned with surrounding assignments; expected 12 spaces but found 13 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
-----------------------------------------------------------------------------------------------

FILE: includes/Sync/Order_Mapper.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 52 | ERROR | A "hook" comment must be a "/**" style docblock comment. (WooCommerce.Commenting.CommentHooks.HookCommentWrongStyle)
-----------------------------------------------------------------------------------------------

FILE: includes/Sync/Order_Polling.php
-----------------------------------------------------------------------------------------------
FOUND 2 ERRORS AND 1 WARNING AFFECTING 3 LINES
-----------------------------------------------------------------------------------------------
  36 | ERROR | Short array syntax is not allowed (Universal.Arrays.DisallowShortArraySyntax.Found)
  68 | WARNING | Equals sign not aligned correctly; expected 1 space but found 2 spaces (Generic.Formatting.MultipleStatementAlignment.IncorrectWarning)
 199 | ERROR | Short array syntax is not allowed (Universal.Arrays.DisallowShortArraySyntax.Found)
-----------------------------------------------------------------------------------------------

FILE: includes/Gateway/API.php
-----------------------------------------------------------------------------------------------
FOUND 6 ERRORS AND 0 WARNINGS AFFECTING 6 LINES
-----------------------------------------------------------------------------------------------
 724 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 725 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 726 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 727 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 729 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 747 | ERROR | The $text parameter must be a single text string literal. Found: 'Failed to make request searchOrders: ' . $response->get_data() (WordPress.WP.I18n.NonSingularStringLiteralText)
-----------------------------------------------------------------------------------------------

FILE: includes/Gateway/API/Requests/Orders.php
-----------------------------------------------------------------------------------------------
FOUND 3 ERRORS AND 0 WARNINGS AFFECTING 3 LINES
-----------------------------------------------------------------------------------------------
 116 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 117 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 118 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
-----------------------------------------------------------------------------------------------
Error: Process completed with exit code 1.

faisal-alvi committed:
a few more phpcs fixes,
FILE: includes/Sync/Order_Polling.php
-----------------------------------------------------------------------------------------------
FOUND 4 ERRORS AND 0 WARNINGS AFFECTING 4 LINES
-----------------------------------------------------------------------------------------------
  82 | ERROR | Expected 1 spaces after opening parenthesis; 0 found (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
  82 | ERROR | Expected 1 spaces before closing parenthesis; 0 found (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
 350 | ERROR | Expected 1 spaces after opening parenthesis; 0 found (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
 350 | ERROR | Expected 1 spaces before closing parenthesis; 0 found (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
-----------------------------------------------------------------------------------------------

FILE: includes/Gateway/API.php
-----------------------------------------------------------------------------------------------
FOUND 3 ERRORS AND 0 WARNINGS AFFECTING 3 LINES
-----------------------------------------------------------------------------------------------
 724 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 726 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 751 | ERROR | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$response'. (WordPress.Security.EscapeOutput.ExceptionNotEscaped)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
Add order note for Square Order ID

faisal-alvi committed:
Update fulfillment UID retrieval to prevent key input append

faisal-alvi committed:
e2e: Add test for order sync and fulfillment confirmation

faisal-alvi committed:
phpcs fixes,
FILE: includes/Gateway.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 498 | ERROR | A function call to esc_html__() with texts containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. (WordPress.WP.I18n.MissingTranslatorsComment)
-----------------------------------------------------------------------------------------------

FILE: includes/Gateway/Cash_App_Pay_Gateway.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 1210 | ERROR | A function call to esc_html__() with texts containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. (WordPress.WP.I18n.MissingTranslatorsComment)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
e2e: Enhance order sync and fulfillment test with Square API integration and order state verification

faisal-alvi committed:
Add checks for shipment and pickup details before updating metadata

faisal-alvi committed:
e2e: increase timeout to fix timeout early

faisal-alvi committed:
phpcs fixes,
FILE: includes/Sync/Order_Importer.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 5 WARNINGS AFFECTING 6 LINES
-----------------------------------------------------------------------------------------------
 155 | WARNING | Equals sign not aligned with surrounding assignments; expected 1 space but found 2 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 156 | WARNING | Equals sign not aligned with surrounding assignments; expected 4 spaces but found 5 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 157 | WARNING | Equals sign not aligned with surrounding assignments; expected 9 spaces but found 10 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 169 | ERROR | Functions must not contain multiple empty lines in a row; found 2 empty lines (Squiz.WhiteSpace.SuperfluousWhitespace.EmptyLines)
 184 | WARNING | Equals sign not aligned with surrounding assignments; expected 5 spaces but found 6 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 185 | WARNING | Equals sign not aligned with surrounding assignments; expected 1 space but found 2 spaces (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
reset last polling time based on seconds_past parameter; useful in testing

faisal-alvi committed:
🌅 Say sayonara to the beta Product Editor,
https://developer.woocommerce.com/2025/07/23/10-1-pre-release-updates/

faisal-alvi committed:
phpcs fix for $_GET param,
FILE: includes/Sync/Order_Polling.php
-----------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 2 LINES
-----------------------------------------------------------------------------------------------
 335 | WARNING | Processing form data without nonce verification. (WordPress.Security.NonceVerification.Recommended)
 336 | WARNING | Processing form data without nonce verification. (WordPress.Security.NonceVerification.Recommended)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
e2e: add product creation step after "Product Block Editor" test removal; that has it created earlier

faisal-alvi committed:
e2e: rename test file after Product Editor related test removal

faisal-alvi committed:
e2e: test needs be entirely removed as Product Block Editor is not supported anymore

faisal-alvi committed:
e2e: update console log messages for clarity in order status verification

faisal-alvi commented:
Thanks @Sidsector9 and @iamdharmesh for the helpful suggestions and questions.

- All of them have been addressed! ✅
- Added an E2E test to cover the implementation
- Removed one outdated test related to the now-removed beta Product Editor, per [WooCommerce 10.1 pre-release update](https://developer.woocommerce.com/2025/07/23/10-1-pre-release-updates/)

Requesting a re-review whenever convenient.

faisal-alvi committed:
Validate square_total directly

faisal-alvi committed:
Remove Lighweight solution for "Shifting Window Problem"; ineffective

faisal-alvi committed:
Enhance order search functionality to prevent Shifting Window Problem

faisal-alvi committed:
phpcs fix; Remove whitespaces

faisal-alvi commented:
@Sidsector9 This is ready for your review following the latest feedback implementation, as noted [here](https://github.com/woocommerce/woocommerce-square/pull/382#discussion_r2260518766).

faisal-alvi committed:
Prevent re-processing of orders by adding 1ms to the start time.,
https://fueled.slack.com/archives/C01UCQGPBDJ/p1754578255593189?thread_ts=1754487914.332999&cid=C01UCQGPBDJ

Sidsector9 reviewed (approved):
Thanks for working on those changes @faisal-alvi !

faisal-alvi committed:
Update order state filter to include CANCELLED status in order search functionality

faisal-alvi committed:
Fix order state filter to use 'CANCELED' instead of 'CANCELLED' for consistency with Square API

faisal-alvi committed:
Merge branch 'trunk' into feat/bidirect-order-sync-woo-to-square

qasumitbagthariya commented:
### QA Update  ✅ 
-------------

I have verified this PR in the `enhancement/341` branch, which has been fixed and is functioning as intended.

I tested the following on this branch:

1. Test WooCommerce to Square Sync: ✅

> - Create a new order in WooCommerce using Square Payment Gateway
> - Verify order appears in Square Dashboard with correct fulfillment details
> - Check fulfillment options are visible at the Square order and match with WooCommerce

[Square Dashboard.pdf](https://github.com/user-attachments/files/21833584/Square.Dashboard.pdf)

<img width="3024" height="3270" alt="image" src="https://github.com/user-attachments/assets/364066d8-7d1f-401f-a3e7-b994e61c0e26" />



2. Test Square to WooCommerce Sync: ✅

> - Update fulfillment status in Square Dashboard (e.g., mark as delivered/shipped)
> - Verify Square order status is now "completed"
> - Verify WooCommerce order status updates accordingly (this may require waiting for 15 minutes or manually running woocommerce_square_poll_orders event; run from the /wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending page)
> - Check WooCommerce order is set to "completed"


https://github.com/user-attachments/assets/96765aed-96cd-4db0-8eb2-734236e4b44b


<img width="1505" height="784" alt="image" src="https://github.com/user-attachments/assets/12c7485d-1ec6-4bac-8455-498e063fe50e" />


3. Test with Different Payment Methods:

> Check the same process for Credit Card, Cash App, and Digital Wallet Payments.
> Check order with Gift Card.


https://github.com/user-attachments/assets/10ce35c3-861a-49e4-a3b2-10e54d3fd897


https://github.com/user-attachments/assets/cfb6a922-4076-4b6e-b4a7-aae620a781da


4. Test Other Cases:

> Confirm Capture Order functionality remains unaffected and continues to work as expected ✅
> Confirm proper error handling for failed syncs 


**Testing Environment** 

<Details>

- WordPress: 6.8.2
- Theme: Storefront 4.6.1
- Theme: Twenty Twenty-Five 1.2
- WooCommerce - 10.0.4
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: feat/bidirect-order-sync-woo-to-square

</Details>


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for UAT
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [x] Yes
- [ ] Not Required/Applicable for this PR
------------------------------------------------	



@faisal-alvi Other than that I notice below points.

1. I noticed some cancel behavior that looks different. When we cancel the order from the Square dashboard, the WooCommerce status should update to 'Cancelled,' but instead, it appears as 'Completed

https://github.com/user-attachments/assets/a9807661-dd77-4dd8-a4e9-3726b3cc7df0

2. Refund not work from the Square board.

https://github.com/user-attachments/assets/f88e3264-1bc3-4be2-85c3-2e98556416d2

https://github.com/user-attachments/assets/5b03002f-b85f-41fa-afd5-8ad0a1c6f1c4

3. I noticed that when sync fails, the error is logged but not displayed in the edit order section. Just wanted to confirm if this is the expected behavior, or should we also display the error in the edit order section?

<img width="1510" height="671" alt="image" src="https://github.com/user-attachments/assets/ddefe8f9-a1ca-4c1b-bee1-146e09c97600" />

<img width="3024" height="3580" alt="image" src="https://github.com/user-attachments/assets/3eac57c9-8363-496b-81f5-b0e71177a9d8" />

faisal-alvi commented:
@qasumitbagthariya Thanks for the testing!

#### 1. Cancel Behavior

When Square line items are marked as Cancelled in the Square Dashboard:
- The line-item fulfillment status changes to CANCELLED.
- However, the overall Square order status changes to COMPLETED.
- On the next sync cycle, WooCommerce reflects this by marking the entire order as Completed, since WooCommerce doesn’t support line-item–level fulfillment statuses, only an overall order status. cc @dkotter

#### 2. Refunds from Square Dashboard

The refund behavior in the Square Dashboard is not in our control and may also be related to using a test card in the sandbox account. Phase 1 is focused only on fulfillment status sync across both platforms, other things, including refund handling, can be considered as part of future enhancements if needed.

#### 3. Error Logging vs Display in Edit Order

During our on-call testing, we added code to intentionally trigger an error:

```
// force error to trigger the log
throw new \Exception( 'Test error' );
```

This was mistakenly placed above:

```
$existing_order = $importer->find_existing_wc_order_by_square_order_id( $order_id );
```

Because of this, `$existing_order` was `null`, which prevented the error note from being added to the order. If placed correctly, the error would have been logged and shown in the Edit Order section as intended.

faisal-alvi committed:
Add opt-in setting for bidirectional order sync

faisal-alvi committed:
add a check for sync settings before syncing orders

faisal-alvi committed:
Implement unscheduling of order sync events and initialize order sync handler conditionally

faisal-alvi committed:
Add method to check if order fulfillment sync is enabled

faisal-alvi committed:
Add constant for order sync event hook definition

faisal-alvi committed:
use defined event hook constant

faisal-alvi committed:
add pre-test setup for order fulfillment sync

faisal-alvi committed:
Link Square Order ID to the Square dashboard in order notes

faisal-alvi committed:
phpcs fixes,
FILE: includes/Admin/Rest/WC_REST_Square_Settings_Controller.php
-----------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 154 | WARNING | Array double arrow not aligned correctly; expected 4 space(s) between "'enable_order_fulfillment_sync'" and double arrow, but found 1. (WordPress.Arrays.MultipleStatementAlignment.DoubleArrowNotAligned)
-----------------------------------------------------------------------------------------------

FILE: includes/Gateway/API/Requests/Orders.php
-----------------------------------------------------------------------------------------------
FOUND 2 ERRORS AND 0 WARNINGS AFFECTING 2 LINES
-----------------------------------------------------------------------------------------------
 676 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 676 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
Merge branch 'trunk' into feat/bidirect-order-sync-woo-to-square

faisal-alvi committed:
Remove branch restriction for PR events in e2e workflow

faisal-alvi committed:
Merge branch 'feat/bidirect-order-sync-woo-to-square' into feat/bidirect-order-sync-opt-in

faisal-alvi committed:
Update version from x.x.x to 4.9.9

faisal-alvi committed:
Merge branch 'feat/bidirect-order-sync-opt-in' of github.com:woocommerce/woocommerce-square into feat/bidirect-order-sync-opt-in

faisal-alvi committed:
Apply suggestions from code review

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

faisal-alvi committed:
Update cash app order notes by linking Square Order ID

dkotter committed:
Merge pull request #396 from woocommerce/feat/bidirect-order-sync-opt-in

Add Opt-in setting for bidirectional order sync

qasumitbagthariya commented:
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.8.4, npm version 10.8.2, node version 20.18.1)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

**Testing Environment** 

<Details>

- WordPress: 6.8.2
- Theme: Twenty Twenty-Five 1.3
- WooCommerce - 10.1.0
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: smoke-testing

</Details>

Next Step- Ready to Merge 🚀
